src/main/java/org/radarcns/consumer/realtime/action/appserver/AppserverClient.java

```diff
@@ -0,0 +1,117 @@
+package org.radarcns.consumer.realtime.action.appserver;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.util.Map;
+import okhttp3.Headers;
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import org.radarbase.exception.TokenException;
+import org.radarbase.oauth.OAuth2Client;
+import org.radarcns.consumer.realtime.action.ActiveAppNotificationAction.MessagingType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Client to provide interface to the AppServer REST API. It uses the OAuth client library from
+ * management portal to authenticate requests.
+ */
+public class AppserverClient {
+
+  private static final Logger logger = LoggerFactory.getLogger(AppserverClient.class);
+
+  private final String baseUrl;
+  private final OkHttpClient httpClient;
+  private final ObjectMapper objectMapper;
+  private final OAuth2Client oauthClient;
+
+  public AppserverClient(String baseUrl, String tokenUrl, String clientId, String clientSecret)
+      throws MalformedURLException {
+    this.baseUrl = baseUrl;
+    httpClient = new OkHttpClient.Builder().build();
+    objectMapper = new ObjectMapper();
+
+    if (tokenUrl == null || tokenUrl.isEmpty()) {
+      logger.warn("MP Token url was not provided. Will use unauthenticated requests to appserver.");
+      this.oauthClient = null;
+    } else {
+      if (clientId.isEmpty()) {
+        throw new IllegalArgumentException("Client ID for appserver client is not set.");
+      }
+      this.oauthClient =
+          new OAuth2Client.Builder()
+              .credentials(clientId, clientSecret)
+              .endpoint(new URL(tokenUrl))
+              .scopes("SUBJECT.READ SUBJECT.UPDATE PROJECT.READ")
+              .httpClient(httpClient)
+              .build();
+    }
+  }
+
+  protected Headers headers() throws TokenException {
+    if (oauthClient != null) {
+      return Headers.of("Authorization", "Bearer " + oauthClient.getValidToken().getAccessToken());
+    } else {
+      return Headers.of();
+    }
+  }
+
+  public Map<String, Object> createMessage(
```

Comment:
you are referring to the create operation that is done on the service that receives the request. Right? I think this should be either sendMessage or addMessage.

Reply:
yes good point